File: rerun-reports/lane-b/072/20260828-065704/results/storage/data-plane/python/streaming/storage-dp-python-blob-event-notifier/python-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/blob_event_notifier/blob_event_handler.py
from __future__ import annotations

import logging
from typing import Any, Protocol
from urllib.parse import unquote

from azure.core.exceptions import HttpResponseError, ResourceNotFoundError
from azure.core.messaging import CloudEvent
from azure.eventgrid import EventGridEvent

StructuredEvent = EventGridEvent | CloudEvent[Any]

_CONTAINER_MARKER = "/containers/"
_BLOB_MARKER = "/blobs/"
_EXPECTED_RACE_ERROR_CODES = {
    "BlobArchived",
    "BlobBeingRehydrated",
    "BlobNotFound",
    "OperationNotAllowedOnArchivedBlob",
}


class BlobService(Protocol):
    def get_blob_client(self, container: str, blob: str) -> Any: ...


class AsyncBlobService(Protocol):
    def get_blob_client(self, container: str, blob: str) -> Any: ...


def parse_blob_subject(subject: str) -> tuple[str, str]:
    container_start = subject.find(_CONTAINER_MARKER)
    blob_start = subject.find(_BLOB_MARKER, container_start + len(_CONTAINER_MARKER))
    if container_start < 0 or blob_start < 0:
        raise ValueError(f"Invalid Blob Storage event subject: {subject!r}")

    container = subject[container_start + len(_CONTAINER_MARKER) : blob_start]
    blob_name = subject[blob_start + len(_BLOB_MARKER) :]
    if not container or not blob_name:
        raise ValueError(f"Invalid Blob Storage event subject: {subject!r}")
    return unquote(container), unquote(blob_name)


def _subject(event: StructuredEvent) -> str:
    subject = event.subject
    if not subject:
        raise ValueError("Blob lifecycle event has no subject")
    return subject


def _property(properties: Any, name: str, default: Any = None) -> Any:
    if isinstance(properties, dict):
        return properties.get(name, default)
    return getattr(properties, name, default)


def _content_type(properties: Any) -> str:
    settings = _property(properties, "content_settings")
    value = _property(settings, "content_type") if settings else None
    return value or "application/octet-stream"


def _access_tier(properties: Any) -> str:
    value = _property(properties, "blob_tier") or _property(properties, "access_tier")
    return str(value or "unknown")


def _is_expected_race(error: HttpResponseError) -> bool:
    return error.status_code in {404, 409} or error.error_code in _EXPECTED_RACE_ERROR_CODES


class BlobEventHandler:
    def __init__(self, blob_service: BlobService, logger: logging.Logger | None = None) -> None:
        self._blob_service = blob_service
        self._logger = logger or logging.getLogger(__name__)

    def handle_created(self, event: StructuredEvent) -> None:
        container, blob_name = parse_blob_subject(_subject(event))
        blob_client = self._blob_service.get_blob_client(container=container, blob=blob_name)
        try:
            downloader = blob_client.download_blob()
            content = downloader.readall()
            properties = downloader.properties
        except ResourceNotFoundError:
            self._logger.warning(
                "Blob %s/%s no longer exists; skipping created event",
                container,
                blob_name,
            )
            return
        except HttpResponseError as error:
            if not _is_expected_race(error):
                raise
            self._logger.warning(
                "Blob %s/%s is unavailable after a lifecycle change (%s)",
                container,
                blob_name,
                error.error_code or error.status_code,
            )
            return

        size = _property(properties, "size", len(content))
        print(
            f"Blob created: name={blob_name}, size={size}, "
            f"content_type={_content_type(properties)}, access_tier={_access_tier(properties)}"
        )

    def handle_deleted(self, event: StructuredEvent) -> None:
        container, blob_name = parse_blob_subject(_subject(event))
        self._logger.info("Blob deleted: %s/%s", container, blob_name)


class AsyncBlobEventHandler:
    def __init__(
        self,
        blob_service: AsyncBlobService,
        logger: logging.Logger | None = None,
    ) -> None:
        self._blob_service = blob_service
        self._logger = logger or logging.getLogger(__name__)

    async def handle_created(self, event: StructuredEvent) -> None:
        container, blob_name = parse_blob_subject(_subject(event))
        blob_client = self._blob_service.get_blob_client(container=container, blob=blob_name)
        try:
            downloader = await blob_client.download_blob()
            content = await downloader.readall()
            properties = downloader.properties
        except ResourceNotFoundError:
            self._logger.warning(
                "Blob %s/%s no longer exists; skipping created event",
                container,
                blob_name,
            )
            return
        except HttpResponseError as error:
            if not _is_expected_race(error):
                raise
            self._logger.warning(
                "Blob %s/%s is unavailable after a lifecycle change (%s)",
                container,
                blob_name,
                error.error_code or error.status_code,
            )
            return

        size = _property(properties, "size", len(content))
        print(
            f"Blob created: name={blob_name}, size={size}, "
            f"content_type={_content_type(properties)}, access_tier={_access_tier(properties)}"
        )

    async def handle_deleted(self, event: StructuredEvent) -> None:
        container, blob_name = parse_blob_subject(_subject(event))
        self._logger.info("Blob deleted: %s/%s", container, blob_name)

File: rerun-reports/lane-a/031/20260828-032156/results/storage/data-plane/java/streaming/storage-dp-java-blob-event-notifier/java-azure-skills/azure-skill-mcp/generated-code/src/main/java/com/example/blobevents/BlobEventHandler.java
package com.example.blobevents;

import com.azure.core.util.BinaryData;
import com.azure.storage.blob.BlobClient;
import com.azure.storage.blob.BlobServiceAsyncClient;
import com.azure.storage.blob.BlobServiceClient;
import com.azure.storage.blob.models.BlobErrorCode;
import com.azure.storage.blob.models.BlobProperties;
import com.azure.storage.blob.models.BlobStorageException;
import java.net.URI;
import java.util.Map;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Mono;

public final class BlobEventHandler {
    public static final String BLOB_CREATED = "Microsoft.Storage.BlobCreated";
    public static final String BLOB_DELETED = "Microsoft.Storage.BlobDeleted";

    private static final Logger LOGGER = LoggerFactory.getLogger(BlobEventHandler.class);
    private static final String CONTAINER_MARKER = "/containers/";
    private static final String BLOB_MARKER = "/blobs/";

    private final BlobServiceClient blobServiceClient;
    private final BlobServiceAsyncClient blobServiceAsyncClient;
    private final Map<BlobAddress, BlobSummary> demoBlobs;

    public BlobEventHandler(
        BlobServiceClient blobServiceClient,
        BlobServiceAsyncClient blobServiceAsyncClient
    ) {
        this.blobServiceClient = Objects.requireNonNull(blobServiceClient, "blobServiceClient");
        this.blobServiceAsyncClient = Objects.requireNonNull(blobServiceAsyncClient, "blobServiceAsyncClient");
        this.demoBlobs = null;
    }

    public static BlobEventHandler forDemo(Map<BlobAddress, BlobSummary> demoBlobs) {
        return new BlobEventHandler(Map.copyOf(demoBlobs));
    }

    private BlobEventHandler(Map<BlobAddress, BlobSummary> demoBlobs) {
        this.blobServiceClient = null;
        this.blobServiceAsyncClient = null;
        this.demoBlobs = demoBlobs;
    }

    public void handle(BlobLifecycleEvent event) {
        switch (event.eventType()) {
            case BLOB_CREATED -> handleCreated(event);
            case BLOB_DELETED -> logDeletion(event);
            default -> LOGGER.warn("Ignoring unrecognized event type '{}' for event {}", event.eventType(), event.id());
        }
    }

    public Mono<Void> handleAsync(BlobLifecycleEvent event) {
        return switch (event.eventType()) {
            case BLOB_CREATED -> handleCreatedAsync(event);
            case BLOB_DELETED -> Mono.fromRunnable(() -> logDeletion(event));
            default -> Mono.fromRunnable(
                () -> LOGGER.warn("Ignoring unrecognized event type '{}' for event {}", event.eventType(), event.id())
            );
        };
    }

    private void handleCreated(BlobLifecycleEvent event) {
        BlobAddress address = parseAddress(event.subject());
        if (demoBlobs != null) {
            printSummary(address, requireDemoBlob(address));
            return;
        }

        BlobClient blobClient = blobServiceClient
            .getBlobContainerClient(address.container())
            .getBlobClient(address.blobName());
        try {
            BlobProperties properties = blobClient.getProperties();
            BinaryData ignoredContent = blobClient.downloadContent();
            printSummary(address, fromProperties(properties));
        } catch (BlobStorageException exception) {
            handleStorageRace(address, exception);
        }
    }

    private Mono<Void> handleCreatedAsync(BlobLifecycleEvent event) {
        BlobAddress address = parseAddress(event.subject());
        if (demoBlobs != null) {
            return Mono.fromRunnable(() -> printSummary(address, requireDemoBlob(address)));
        }

        return blobServiceAsyncClient
            .getBlobContainerAsyncClient(address.container())
            .getBlobAsyncClient(address.blobName())
            .getProperties()
            .flatMap(properties -> blobServiceAsyncClient
                .getBlobContainerAsyncClient(address.container())
                .getBlobAsyncClient(address.blobName())
                .downloadContent()
                .doOnNext(ignored -> printSummary(address, fromProperties(properties))))
            .onErrorResume(BlobStorageException.class, exception -> {
                handleStorageRace(address, exception);
                return Mono.empty();
            })
            .then();
    }

    private static void logDeletion(BlobLifecycleEvent event) {
        BlobAddress address = parseAddress(event.subject());
        LOGGER.info("Blob deleted: container='{}', name='{}'", address.container(), address.blobName());
    }

    private static BlobSummary fromProperties(BlobProperties properties) {
        String contentType = properties.getContentType() == null ? "application/octet-stream" : properties.getContentType();
        String accessTier = properties.getAccessTier() == null ? "unknown" : properties.getAccessTier().toString();
        return new BlobSummary(properties.getBlobSize(), contentType, accessTier);
    }

    private static void printSummary(BlobAddress address, BlobSummary summary) {
        LOGGER.info(
            "Blob created: name='{}', container='{}', size={} bytes, contentType='{}', accessTier='{}'",
            address.blobName(),
            address.container(),
            summary.size(),
            summary.contentType(),
            summary.accessTier()
        );
    }

    private BlobSummary requireDemoBlob(BlobAddress address) {
        BlobSummary summary = demoBlobs.get(address);
        if (summary == null) {
            LOGGER.warn("Blob disappeared before it could be downloaded: {}/{}", address.container(), address.blobName());
            return new BlobSummary(0, "unknown", "unknown");
        }
        return summary;
    }

    private static void handleStorageRace(BlobAddress address, BlobStorageException exception) {
        if (exception.getStatusCode() == 404) {
            LOGGER.warn("Blob disappeared before it could be downloaded: {}/{}", address.container(), address.blobName());
            return;
        }
        if (BlobErrorCode.BLOB_ARCHIVED.equals(exception.getErrorCode())
            || BlobErrorCode.BLOB_BEING_REHYDRATED.equals(exception.getErrorCode())) {
            LOGGER.warn(
                "Blob is currently unavailable, possibly because its access tier changed: {}/{} ({})",
                address.container(),
                address.blobName(),
                exception.getErrorCode()
            );
            return;
        }
        throw exception;
    }

    public static BlobAddress parseAddress(String subject) {
        int containerStart = subject.indexOf(CONTAINER_MARKER);
        int blobStart = subject.indexOf(BLOB_MARKER, containerStart + CONTAINER_MARKER.length());
        if (containerStart < 0 || blobStart < 0) {
            throw new IllegalArgumentException("Blob event subject has an unexpected format: " + subject);
        }
        String container = subject.substring(containerStart + CONTAINER_MARKER.length(), blobStart);
        String encodedBlobName = subject.substring(blobStart + BLOB_MARKER.length());
        String blobName = URI.create(encodedBlobName).getPath();
        if (container.isBlank() || blobName.isBlank()) {
            throw new IllegalArgumentException("Blob event subject must identify a container and blob: " + subject);
        }
        return new BlobAddress(container, blobName);
    }

    public record BlobAddress(String container, String blobName) {
    }

    public record BlobSummary(long size, String contentType, String accessTier) {
    }
}
